### PR TITLE
Add Telegram support

### LINKED ISSUE #736: Can you add  Telegram alerts to bot?
Can you add  Telegram alerts to bot?
Because Slack is not  used everywhere and by every one :D

### PR DESCRIPTION
**Pull request by yaodingyd**
# What did you implement:

Fixes #736 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

# Checklist:
You don't have to satisfy all of the following.

- [x] Write tests
- [x] Write documentation
- [x] Check that there aren't other open pull requests for the same issue/feature
- [x] Format your source code by `make fmt`
- [x] Pass the test by `make test`
- [x] Provide verification config / commands
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** Yes  

# Reference

* https://blog.github.com/2015-01-21-how-to-write-the-perfect-pull-request/

image from ios

### CODE REVIEW

File: report/telegram.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/future-architect/vuls/config"
+	"github.com/future-architect/vuls/models"
+)
+
+// ChatWorkWriter send report to ChatWork
```
Comment (kotakanbe) [inline report/telegram.go]: __[Golint]__ comment on exported type TelegramWriter should be of the form "TelegramWriter ..." (with optional leading article)

<sub>posted by Sider</sub>

File: report/telegram.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/future-architect/vuls/config"
+	"github.com/future-architect/vuls/models"
+)
+
+type TelegramWriter struct{}
```
Comment (kotakanbe) [inline report/telegram.go]: __[Golint]__ exported type TelegramWriter should have comment or be unexported

<sub>posted by Sider</sub>

**Comment by kotakanbe** [comment]
@yaodingyd 
Many thanks for sending P/R :)
Would you please fix the doc? 

https://github.com/vulsdoc/vuls
https://github.com/vulsdoc/vuls/blob/master/docs/usage-report.md

**Comment by kotakanbe** [comment]
@yaodingyd Thanks 👍 

Would you add a template of `telegram` to discover.go?
https://github.com/future-architect/vuls/blob/master/commands/discover.go#L178

Executing this command with 127.0.0.1 is useful when displaying the template of config.toml.
`$ vuls discover 127.0.0.1/32`

**Comment by kotakanbe** [comment]
Memo: 
Markdown style: https://core.telegram.org/bots/api#markdown-style

**Comment by kotakanbe** [comment]
@yaodingyd 

I tried, but ` Status:     "429 Too Many Requests",` Returned.
We have to reduce the number of Requests. 
Your P/R sends ScannedCVEs one by one, but it seems necessary to send more than one at once.

**Comment by yaodingyd** [comment]
@kotakanbe I have grouped message by 10 to send

File: config/config.go [hunk 3]
```diff
@@ -557,6 +563,32 @@ func (c *ChatWorkConf) Validate() (errs []error) {
 	return
 }
 
+// TelegramConf is Telegram config
+type TelegramConf struct {
+	Token   string `json:"-"`
+	Channel string `json:"-"`
```
Comment (kotakanbe) [inline config/config.go]: ChatID is better.
https://core.telegram.org/bots/api#sendmessage

File: report/telegram.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			}
+		}
+
+	}
+	return nil
+}
+
+func sendMessage(channel, token, message string) error {
+	uri := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", token)
+	payload := `{"text": "` + message + `", "chat_id": "@` + channel + `", "parse_mode": "Markdown" }`
```
Comment (kotakanbe) [inline report/telegram.go]: remove '@' of chat_id.
'@' is no need when sending a message to the user.
see https://stackoverflow.com/questions/31078710/how-to-obtain-telegram-chat-id-for-a-specific-user

File: config/config.go (hunk 3 above)
Comment (kotakanbe) [reply config/config.go]: Please replace Channel to ChatID in your P/R.

**Comment by kotakanbe** [comment]
@yaodingyd 

Fixed some errors.  Please apply this diff to your P/R and please test by yourself.

- escape " to \" in json body.
- Fix `message is too long` error. (change 10 to 5) 
- forget to reset the counter.
- some refactoring

```
diff --git i/report/telegram.go w/report/telegram.go
index 7a19596..eaf1dcd 100644
--- i/report/telegram.go
+++ w/report/telegram.go
@@ -18,33 +18,35 @@ func (w TelegramWriter) Write(rs ...models.ScanResult) (err error) {
        conf := config.Conf.Telegram

        for _, r := range rs {
-               serverInfo := fmt.Sprintf("%s", r.ServerInfo())
-               counter := 0
-               message := ""
+               msgs := []string{fmt.Sprintf("*%s*\n%s\n%s\n%s",
+                       r.ServerInfo(),
+                       r.ScannedCves.FormatCveSummary(),
+                       r.ScannedCves.FormatFixedStatus(r.Packages),
+                       r.FormatUpdatablePacksSummary())}
+
                for _, vinfo := range r.ScannedCves {
-                       counter++
                        maxCvss := vinfo.MaxCvssScore()
                        severity := strings.ToUpper(maxCvss.Value.Severity)
                        if severity == "" {
                                severity = "?"
                        }

-                       message += fmt.Sprintf(`*%s* \n [%s](https://nvd.nist.gov/vuln/detail/%s) _%s %s_\n %s\n\n`,
-                               serverInfo,
+                       msgs = append(msgs, fmt.Sprintf(`[%s](https://nvd.nist.gov/vuln/detail/%s) _%s %s %s_\n %s`,
                                vinfo.CveID,
                                vinfo.CveID,
                                strconv.FormatFloat(maxCvss.Value.Score, 'f', 1, 64),
                                severity,
-                               vinfo.Summaries(config.Conf.Lang, r.Family)[0].Value)
-                       if counter == 10 {
-                               message = ""
-                               if err = sendMessage(conf.ChatID, conf.Token, message); err != nil {
+                               maxCvss.Value.Vector,
+                               vinfo.Summaries(config.Conf.Lang, r.Family)[0].Value))
+                       if len(msgs) == 5 {
+                               if err = sendMessage(conf.ChatID, conf.Token, strings.Join(msgs, "\n\n")); err != nil {
                                        return err
                                }
+                               msgs = []string{}
                        }
                }
-               if message != "" {
-                       if err = sendMessage(conf.ChatID, conf.Token, message); err != nil {
+               if len(msgs) != 0 {
+                       if err = sendMessage(conf.ChatID, conf.Token, strings.Join(msgs, "\n\n")); err != nil {
                                return err
                        }
                }
@@ -55,23 +57,19 @@ func (w TelegramWriter) Write(rs ...models.ScanResult) (err error) {

 func sendMessage(chatID, token, message string) error {
        uri := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", token)
-       payload := `{"text": "` + message + `", "chat_id": "` + chatID + `", "parse_mode": "Markdown" }`
-
+       payload := `{"text": "` + strings.Replace(message, `"`, `\"`, -1) + `", "chat_id": "` + chatID + `", "parse_mode": "Markdown" }`
        req, err := http.NewRequest("POST", uri, bytes.NewBuffer([]byte(payload)))
        req.Header.Add("Content-Type", "application/json")
-
        if err != nil {
                return err
        }
        client := &http.Client{}
-
        resp, err := client.Do(req)
        if checkResponse(resp) != nil && err != nil {
                fmt.Println(err)
                return err
        }
        defer resp.Body.Close()
-
        return nil
 }
```

**Comment by yaodingyd** [comment]
@kotakanbe refactor applied. Maybe next time you can directly work on my PR? I feel like applying diff is error-prone...

**Comment by kotakanbe** [comment]
@yaodingyd Many Thanks.

Get $50 👍 
https://issuehunt.io/repos/54808920/issues/736

**Comment by yaodingyd** [comment]
@kotakanbe Thanks! Already submitted.

**Pull request merged**